### PR TITLE
Add feature flag to disable Fiber-only features

### PR DESCRIPTION
**Pull request by acdlite**
Disables Fiber-only render return types

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 1]
```diff
@@ -101,9 +102,10 @@ module.exports = function<T, P, I, TI, PI, C, CX>(config : HostConfig<T, P, I, T
 
   // The priority level to use when scheduling an update.
   // TODO: Should we change this to an array? Might be less confusing.
-  let priorityContext : PriorityLevel = useSyncScheduling ?
-    SynchronousPriority :
-    LowPriority;
+  let priorityContext : PriorityLevel =
+    (useSyncScheduling || ReactFeatureFlags.disableNewFiberFeatures) ?
```
Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: I'm fine leaving this as before too – either way.

File: src/renderers/shared/fiber/ReactFiberReconciler.js [hunk 2]
```diff
@@ -134,6 +138,11 @@ module.exports = function<T, P, I, TI, PI, C, CX>(config : HostConfig<T, P, I, T
     },
 
     updateContainer(element : ReactNodeList, container : OpaqueNode, parentComponent : ?ReactComponent<any, any, any>, callback: ?Function) : void {
+      invariant(
+        !ReactFeatureFlags.disableNewFiberFeatures || isValidElement(element),
+        'render(): Invalid component element.'
```
Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberReconciler.js]: Can you make sure null/false works? And we might as well use the old error message:

```
      invariant(
        inst === null ||
        inst === false ||
        React.isValidElement(inst),
        '%s(...): A valid React element (or null) must be returned. You may have ' +
        'returned undefined, an array or some other invalid object.',
        Component.displayName || Component.name || 'Component'
      );
```

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberReconciler.js]: I think that's the message for when you return a non-element in a composite component. The one for top-level is here: https://github.com/facebook/react/blob/master/src/renderers/dom/stack/client/ReactMount.js#L439

File: src/renderers/shared/fiber/ReactChildFiber.js [hunk 3]
```diff
@@ -1100,6 +1101,41 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
...
+              currentFirstChild,
+              newChild,
+              priority
+            ));
+        }
+      }
+
+      if (__DEV__) {
+        const Component = returnFiber.type;
+        warning(
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactChildFiber.js]: Oops this should be an invariant

Comment (acdlite) [inline src/renderers/shared/fiber/ReactChildFiber.js]: Well huh, it looks like sometimes Stack warns and sometimes it's an invariant?

https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L46

Comment (gaearon) [inline src/renderers/shared/fiber/ReactChildFiber.js]: Stack *both* warns and does an invariant. This is an artifact from some earlier refactor when this was a different warning. See https://github.com/facebook/react/pull/8648/commits/2646272e8ac55a15bbfde0dc49b0e5294b7cd5fd and feel free to reuse my other commits in #8648.

File: src/renderers/shared/fiber/ReactFiberReconciler.js (hunk 2 above)
Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberReconciler.js]: Ahh yes sorry.

File: src/renderers/shared/fiber/ReactChildFiber.js (hunk 3 above)
Comment (spicyj) [inline src/renderers/shared/fiber/ReactChildFiber.js]: Occasionally we do that intentionally too if the warning gives a better stack.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactChildFiber.js]: Haha didn't see that PR, sorry Dan! I'll rebase on top

Comment (gaearon) [inline src/renderers/shared/fiber/ReactChildFiber.js]: >Occasionally we do that intentionally too if the warning gives a better stack.

Yep, but isn’t the case there. (Both stacks are internal and roughly in the same place.)

>I'll rebase on top

You probably only want da3dc0354dc37af431f59ac04a29e694515511cd, 243d24542354eae55a77ce68ea556e04a13967be, 2646272e8ac55a15bbfde0dc49b0e5294b7cd5fd.

I haven't looked at your PR yet but you might want to do 8c4b2a64dbb848bd6659821b0c3af771f283e03c and 5643a2c6272ca27cff5c4b739638cce266ee1dc6 differently.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactChildFiber.js]: I'll cherry-pick everything except the last commit, in favor of the way I did it here, for the reason Sebastian gave here: https://github.com/facebook/react/pull/8648#discussion_r94506177

File: src/renderers/dom/fiber/__tests__/ReactDOMFiber-test.js [hunk 4]
```diff
@@ -1002,5 +1002,50 @@ describe('ReactDOMFiber', () => {
         container
       );
     });
+
+    describe('disableNewFiberFeatures', () => {
+      var ReactFeatureFlags = require('ReactFeatureFlags');
```
Comment (sptq) [inline src/renderers/dom/fiber/__tests__/ReactDOMFiber-test.js]: You could change `var` to `const`

File: src/renderers/shared/fiber/ReactFiberReconciler.js [hunk 5]
```diff
@@ -17,6 +17,10 @@ import type { FiberRoot } from 'ReactFiberRoot';
 import type { PriorityLevel } from 'ReactPriorityLevel';
 import type { ReactNodeList } from 'ReactTypes';
 
+var { isValidElement } = require('ReactElement');
+var invariant = require('invariant');
+var ReactFeatureFlags = require('ReactFeatureFlags');
+
```
Comment (sptq) [inline src/renderers/shared/fiber/ReactFiberReconciler.js]: You can change those to `import` instead of `require`

Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberReconciler.js]: Actually, not all our tooling works with ES6 imports yet so we can't. require() works fine for us for now.

File: src/renderers/dom/fiber/__tests__/ReactDOMFiber-test.js (hunk 4 above)
Comment (acdlite) [inline src/renderers/dom/fiber/__tests__/ReactDOMFiber-test.js]: We're already pretty inconsistent throughout the codebase :D One day we'll go through and clean this up

**Pull request merged**